Add 'etesync' Ansible role

ypid opened the pull request:
The first new DebOps role of the decade :)

Project: https://www.etesync.com/

WIP: @drybjed I am not sure if the integration test is ok. I have not yet played with the current testing approach enough.

Implements: #1233 
Updates: #1230

Reasons to update #1230:

* The etckeeper role should not get bloated by each and every application that could possibly write secret files to /etc.
* Users might not use the DebOps etckeeper role and instead only use the application roles and would then not have secret files excluded from Etckeeper if they still use it.

ypid committed:
Add 'etesync' Ansible role

Implements: https://github.com/debops/debops/issues/1233

ypid committed:
[etesync] Move .gitignore from the etckeeper role to this one

Reasons:

* The etckeeper role should not get bloated by each and every
  application that could possibly write secret files to /etc.
* Users might not use the DebOps etckeeper role and instead only use the
  application roles and would then not have secret files excluded from
  Etckeeper if they still use it.

Updates: https://github.com/debops/debops/pull/1230

ypid committed:
[etesync] More strict file owner/permissions for files below /etc

ypid committed:
[etesync] Fix copy/paste mistake

ypid committed:
[etesync] Add first integration test

SuperSandro2000 commented on ansible/roles/etesync/defaults/main.yml:
```diff
@@ -0,0 +1,456 @@
+  - 'python3-dev'
+
+  ## Django>=2.1.7,<2.1.999 is required which is currently only in Debian sid.
+  # - 'python3-django'
```
Apt pinning?

ypid replied:
Please refer to https://docs.debops.org/en/master/dep/dep-0003.html Debian sid is not enabled on production servers.

SuperSandro2000 replied:
oh yeah. Also DEP-11 is probably ambiguous with debian appstream.

drybjed replied:
Does that mean that EteSync cannot be used on Debian Buster without Sid package installed? Or do you install Django in the virtualenv instead?

~~Edit (ypid): I somehow cannot comment on this on GitHub so I edit instead. For now virtualenv is used.~~

ypid replied:
Sid packages are not required obviously. I just tried to install as much of the dependencies from stable but then had to give up. I use virtualenv and pip for now.

drybjed commented on CHANGELOG.rst:
```diff
@@ -27,6 +27,8 @@ New DebOps roles
 - The :ref:`debops.lxd` role brings support for LXD on Debian hosts by building
   the Go binaries from source, without Snap installation.
 
+- The :ref:`debops.etesync` role allows to setup a EteSync server.
```
Perhaps you could add a sentence or two about what EteSync server can be used for because I'm not sure if it's essential or not... :-) An anonymous link to the website here wouldn't hurt either - this will be a first place new users will hear about it.

ypid replied:
Sounds good. I am not creative so I just copied the intro from the upstream website one more time :-/.

drybjed commented on .gitlab-ci.yml:
```diff
@@ -511,6 +511,15 @@ stages:
     JANE_LOG_PATTERN: '\[debops\.etherpad\]'
 
+'etesync role':
+  <<: *test_role_1st_deps
```
The `nginx` role already is on the first level of dependencies, so this should be on the second level, `test_role_2nd_deps`.

ypid replied:
Thanks, fixed.

drybjed commented on ansible/roles/etckeeper/defaults/main.yml:
```diff
@@ -176,9 +176,6 @@ etckeeper__default_gitignore:
     ignore: 'zfs/zpool.cache'
 
-  - name: 'etesync-secret'
-    ignore: 'etesync-server/secret.txt'
```
At the moment this change is bundled with a change in the `etesync` role. This will be hard to backport cleanly to older DebOps releases. But I suppose that I can just do a separate revert of the https://github.com/debops/debops/commit/c6b19c45fd18523186a6ac3fac5b1da8481839f2 commits.

ypid replied:
Should be fine to revert as you have not made a 1.2.x release yet including it. Then I guess you support this switch in how we use the etckeeper role and my reasoning?

ypid replied:
@drybjed Ping

drybjed replied:
I remember about this. I'll revert the commit when the `etesync` role is merged.

drybjed commented on ansible/roles/etesync/defaults/main.yml:
```diff
@@ -0,0 +1,456 @@
+# git commit hash lock to release 0.2.1.
+# Note that this hash locking is not very effective because the main
+# implementation of EteSync is in additional Python packages.
+etesync__git_version: 'cf0dc8e6a81f0b87bfba6d952ce9085d21257c18'
```
Perhaps you should contact upstream author and ask him to sign tags instead of only random commits? That would make a separate task to verify the signature unnecessary and would let us use git tags instead of specific commit here.

ypid replied:
Upstream really cares about security in my eyes so they already sign tags (ref: https://github.com/etesync/server/tags). When you look at the role, you will notice that I of course also verify this signature.

It is just that over at https://github.com/hashbang/os, we really like the concept of "hash locking" particulare versions even if they are already signed. In Android, it is not unseen that signed tags are changed (not even a compromise, just regular behavior) … That is something where hash locking helps. I don’t see the big downside of having that additional security. When I hash lock some upstream project that usually also indicates that I have done at least some degree of research and review of it and that is best signaled with exactly referring to what version I checked.

drybjed commented on ansible/roles/etesync/defaults/main.yml:
```diff
@@ -0,0 +1,456 @@
+# .. envvar:: etesync__python__dependent_packages3 [[[
+#
+# Configuration for the :ref:`debops.python` Ansible role.
+etesync__python__dependent_packages3:
```
Can EteSync be used with Python 2.7 environment?

ypid replied:
I don’t care about Python2 as it is EOL. Python 3 is ~~installed~~ available everywhere and there it works. Do we still need to care?

> Or do you install Django in the virtualenv instead?

Yes, I have to.

drybjed commented on ansible/roles/etesync/tasks/main.yml:
```diff
@@ -0,0 +1,204 @@
+  debug:
+    msg: 'URL: {{ etesync__url }}'
+  when: etesync__http_psk_subpath != ""
+  tags: [ 'role::etesync:show_url' ]
```
I don't like to have `debug` modules in the production code, this implies that Ansible output is meant to be read and interactive which is not strictly true. We need to find a better, general purpose solution for this. Perhaps a mail to `root` on first execution would suffice? This would also ensure that the message with the details can be archived if necessary. The SMTP configuration should make sure that it will be sent to the correct person.

ypid replied:
Very good idea, thanks. I don’t operate (with DebOps/Ansible) at such a scale but I can definitely see this come in handy. Done.

drybjed commented on docs/ansible/roles/etesync/getting-started.rst:
```diff
@@ -0,0 +1,48 @@
+Example inventory
+-----------------
+
+To configure Docker on a given remote host, it needs to be added to the
```
Docker? What?

drybjed commented on docs/ansible/roles/etesync/getting-started.rst:
```diff
@@ -0,0 +1,48 @@
+Example playbook
+----------------
+
+Here's an example playbook that can be used to manage Docker:
```
And Docker again here. :)

drybjed reviewed (changes requested):
EteSync seems like an interesting solution for the CalDAV/CardDAV conundrum. I wonder if there are any blockers for it to be added to Debian archive, at least the server part.

There are some questions or issues I'd like to see your answers to, before merging this PR. See the comments below.

ypid committed:
[etesync] Add introduction from website with link to the changelog

drybjed committed:
[etesync] Code review: Fix CI test

> The nginx role already is on the first level of dependencies, so this
> should be on the second level, test_role_2nd_deps.

ypid committed:
[etesync] Update to version 0.2.2

drybjed committed:
[docs] Code review: Fix copy/paste mistake

ypid committed:
[etesync] Send email with PSK subpath URL instead of Ansible debug task

Role model of this commit is apparmor where I have done something like
this before.

ypid reviewed (commented):
> EteSync seems like an interesting solution for the CalDAV/CardDAV conundrum. I wonder if there are any blockers for it to be added to Debian archive, at least the server part.

From having reviewed (a bit) the software myself, I could not find any blockers. Most if not all dependencies are already in Debian. I guess the only issue is that nobody has stepped up to package it yet.

> There are some questions or issues I'd like to see your answers to, before merging this PR. See the comments below.

Thanks for your careful (as always) review!

drybjed commented on ansible/roles/etesync/templates/etc/etesync-server/etesync-server.ini.j2:
```diff
@@ -0,0 +1,16 @@
+
+[database]
+engine = django.db.backends.sqlite3
+name = {{ etesync__data }}/db.sqlite3
```
I'm suprised that the role does not support PostgreSQL as an alternative to SQLite. I suppose that this can be added later and backported to a stable release.

ypid replied:
Jup. The only reasons I did not add PostgreSQL support is that for my small deployment I decided to go with SQLite. If someone needs PostgreSQL, it can be added.

SuperSandro2000 replied:
I would like to request postgres as my backup solution can't really deal with sqlite.

drybjed replied:
In that case I suggest that the PR will be merged after `v2.0.0` release which is imminent (perhaps today). I'll revert the `etckeeper` change as well.

ypid replied:
@SuperSandro2000 Feel free to add PostgreSQL support.

drybjed commented on ansible/roles/etesync/tasks/main.yml:
```diff
@@ -0,0 +1,209 @@
+  become: True
+  become_user: '{{ etesync__user }}'
+  when: (not etesync__register_installed.stat.exists|bool and
+         etesync__register_migration is not changed)
```
The above task which is supposed to create a superuser account did not get executed, therefore there's no admin account or password. I suspect that the `when:` conditions are wrong, because the they depend on the database being not changed, which isn't the case on first install. Perhaps you should change the condition. I wonder if the Django command is idempotent or can it be made so if not.

ypid replied:
Good catch, fixed. I copied most of the Django setup from the NetBox role :)

drybjed reviewed (changes requested):
I tried out the EteSync role in a development environment. Application installed fine but `gunicorn` didn't start properly because it's not present on Buster managed by DebOps by default  (Python3-only), but `gunicorn3` is. This is a bug in the `gunicorn` role, I'll fix that shortly.

When I went to the EteSync site I was greeted with message:
```
It works!
Please refer to the README to complete the final steps if you haven't done so already.
```
Is that expected? I thought that was the webserver at first, but it seems that this is indeed the EteSync application. I figured that going to the `/admin` page would give me a login form, but due to superuser account issues mentioned elsewhere I didn't had proper credentials to test it further.

Perhaps you could mention this in the Getting Started section of the role documentation so that it is expected?

ypid committed:
[etesync] Fix superuser account on first run

ypid committed:
[etesync] Add to docs what the current feature set of the role is

ypid commented:
> This is a bug in the gunicorn role, I'll fix that shortly.

Thanks!

> Is that expected?
> Perhaps you could mention this in the Getting Started section of the role documentation so that it is expected?

It is expected. I added a paragraph to the docs about the feature set. I know upstream docs are not comprehensive but I don’t want to document in a role that setups up the service how to use the service itself.

drybjed commented on ansible/roles/etesync/tasks/main.yml:
```diff
@@ -0,0 +1,203 @@
+  become: True
+  become_user: '{{ etesync__user }}'
+  when: ("etesync" not in ansible_local)
+  no_log: True
```
After the change the task is executed, but (on Debian Buster) it errors out with:
```
fatal: [etesync]: FAILED! => {"changed": true, "cmd": "## -o nounset can be enabled with virtualenv >= 16.2.\n## Ref: https://github.com/pypa/virtualenv/pull/922\nset -o pipefail -o errexit;\nsource /usr/local/lib/etesync/virtualenv/bin/activate\necho \"from django.contrib.auth.models import User;\n      User.objects.create_superuser('${SUPERUSER_NAME}', '${SUPERUSER_EMAIL}', '${SUPERUSER_PASSWORD}')\" | ./manage.py shell\n", "delta": "0:00:00.182531", "end": "2020-01-29 22:54:04.625757", "msg": "non-zero return code", "rc": 1, "start": "2020-01-29 22:54:04.443226", "stderr": "Traceback (most recent call last):\n  File \"./manage.py\", line 22, in <module>\n    execute_from_command_line(sys.argv)\n  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/__init__.py\", line 381, in execute_from_command_line\n    utility.execute()\n  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/__init__.py\", line 375, in execute\n    self.fetch_command(subcommand).run_from_argv(self.argv)\n  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/base.py\", line 316, in run_from_argv\n    self.execute(*args, **cmd_options)\n  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/base.py\", line 353, in execute\n    output = self.handle(*args, **options)\n  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/commands/shell.py\", line 92, in handle\n    exec(sys.stdin.read())\n  File \"<string>\", line 2\n    User.objects.create_superuser('drybjed', 'drybjed', 'fE0hheO78pC16Ths.U2Z')\n    ^\nIndentationError: unexpected indent", "stderr_lines": ["Traceback (most recent call last):", "  File \"./manage.py\", line 22, in <module>", "    execute_from_command_line(sys.argv)", "  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/__init__.py\", line 381, in execute_from_command_line", "    utility.execute()", "  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/__init__.py\", line 375, in execute", "    self.fetch_command(subcommand).run_from_argv(self.argv)", "  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/base.py\", line 316, in run_from_argv", "    self.execute(*args, **cmd_options)", "  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/base.py\", line 353, in execute", "    output = self.handle(*args, **options)", "  File \"/usr/local/lib/etesync/virtualenv/lib/python3.7/site-packages/django/core/management/commands/shell.py\", line 92, in handle", "    exec(sys.stdin.read())", "  File \"<string>\", line 2", "    User.objects.create_superuser('drybjed', 'drybjed', 'fE0hheO78pC16Ths.U2Z')", "    ^", "IndentationError: unexpected indent"], "stdout": "", "stdout_lines": []}
```
I think that making it closer to the similar task in the `netbox` role by changing the content to not use the `: |` syntax will fix this, but it's getting late here so I'm heading to bed. I'll try and fix it tommorow if @ypid won't do it first. :-)

ypid replied:
Thanks for the review and the hint. Fixed it.

ypid committed:
[etesync] Fix Python indent issue in superuser setup task

drybjed reviewed (approved):
@ypid Thanks for the new role. Since we are at the beginning of the new DebOps development cycle, I think that the current PR can be emrged and the PostgreSQL support can be added later.

There's a web client in the works, do you know how far along it is from a stable release? It might be useful to have a role for it as well, or add support to the `etesync` role itself in the future.

ypid commented:
You are welcome, thanks for maintaining :+1: The web client is useful and I have tested it a bit. That would be interesting to add to DebOps. When that is done, also [Signed Pages](https://github.com/tasn/webext-signed-pages) should be supported, not sure yet how that can be implemented. I have some interesting in that. But this is not a priority for me meaning that when current demand does not change, I will not setup the web client at all. PostgreSQL I will only do when I run into limits with SQLite in my small deployment.
